reconstructSPI/iterative_refinement/expectation_maximization.py

```diff
@@ -0,0 +1,539 @@
+
+    References
+    ----------
+    1. Nelson, P. C. (2021). Physical Models of Living Systems new
+    chapter: Single Particle Reconstruction in Cryo-electron
+    Microscopy.
+            https://repository.upenn.edu/physics_papers/656
+    2. Scheres, S. H. W. (2012). RELION: Implementation of a
+    Bayesian approach to cryo-EM structure determination.
+            Journal of Structural Biology, 180(3), 519–530.
+            http://doi.org/10.1016/j.jsb.2012.09.006
+    3. Sigworth, F. J., Doerschuk, P. C., Carazo, J.-M., & Scheres,
+    S. H. W. (2010).
+            An Introduction to Maximum-Likelihood Methods in Cryo-EM.
+            In Methods in Enzymology (1st ed., Vol. 482,
+            pp. 263–294). Elsevier Inc.
+            http://doi.org/10.1016/S0076-6879(10)82011-7
+    """
+
+    def __init__(self, map_3d_init, particles, ctf_info, max_itr=7):
+        self.map_3d_init = map_3d_init
+        self.particles = particles
+        self.ctf_info = ctf_info
+        self.max_itr = max_itr
+
+    def iterative_refinement(self, wiener_small_number=0.01, count_norm_const=1):
+        """Perform iterative refinement.
+
+        Acts in a Bayesian expectation maximization setting,
+        i.e. maximum a posteriori estimation.
+
+        Parameters
+        ----------
+        wiener_small_number : float
+            Used to tune Wiener filter.
+        count_norm_const : float
+            Used to tune normalization of slice inserting.
+
+        Returns
+        -------
+        map_3d_update : arr
+            Current iteration of map.
+            Shape (n_pix, n_pix, n_pix)
+        map_3d_final : arr
+            Final updated map.
+            Shape (n_pix, n_pix, n_pix)
+        half_map_3d_final_1 : arr
+            Shape (n_pix, n_pix, n_pix)
+        half_map_3d_final_2 : arr
+            Shape (n_pix, n_pix, n_pix)
+        fsc_1d : arr
+            Final one dimensional fourier shell correlation.
+            Shape (n_pix // 2,)
+        """
+        particles_1, particles_2 = IterativeRefinement.split_array(self.particles)
+
+        ctfs = self.build_ctf_array()
+        ctfs_1, ctfs_2 = IterativeRefinement.split_array(ctfs)
+
+        # work in Fourier space. So particles can stay in Fourier
+        # space the whole time. They are experimental measurements
+        # and are fixed in the algorithm
+        particles_f_1 = IterativeRefinement.fft_3d(particles_1)
+        particles_f_2 = IterativeRefinement.fft_3d(particles_2)
+
+        n_pix = self.map_3d_init.shape[0]
+
+        n_rotations = self.particles.shape[0]
+
+        # suggest 32 or 64 to start with. real data will be more like
+        # 128 or 256. Can have issues with ctf at small pixels and
+        # need to zero pad to avoid artefacts. Artefacts from ctf not
+        # going to zero at edges, and sinusoidal ctf rippling too fast
+        # can zero pad when do Fourier convolution (fft is on zero
+        # padded and larger sized array)
```

Comment:
made an issue for this. can remove comments https://github.com/compSPI/reconstructSPI/issues/29